### ISSUE TITLE
plugin/forward: Panic while healthchecking

### ISSUE THREAD

**Issue by negz** [coredns/coredns#2003]
I encountered the following panic today while experimenting with deploying CoreDNS

```
$ kubectl --context tfk-uc14 -n kube-system logs coredns-7b58f898db-zzttn --previous
.:53
2018/07/25 23:17:32 [INFO] CoreDNS-1.2.0
2018/07/25 23:17:32 [INFO] linux/amd64, go1.10.3, 2e322f6
CoreDNS-1.2.0
linux/amd64, go1.10.3, 2e322f6
2018/07/25 23:17:32 [INFO] plugin/reload: Running configuration MD5 = df9eeab6ca5daae007e09be2306a3d29
panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x18 pc=0x11dd3b9]
goroutine 3941 [running]:
github.com/coredns/coredns/plugin/forward.(*Proxy).Healthcheck.func1(0xc4201b0cd0, 0x0)
        /tmp/tmp.RJbV5C8cVr/g/src/github.com/coredns/coredns/plugin/forward/proxy.go:54 +0x29
github.com/coredns/coredns/plugin/pkg/up.(*Probe).Do.func1(0xc42035d6d0, 0x1dcd6500, 0xc4204d2000)
        /tmp/tmp.RJbV5C8cVr/g/src/github.com/coredns/coredns/plugin/pkg/up/up.go:38 +0x36
created by github.com/coredns/coredns/plugin/pkg/up.(*Probe).Do
        /tmp/tmp.RJbV5C8cVr/g/src/github.com/coredns/coredns/plugin/pkg/up/up.go:36 +0xa2
```

My `Corefile` was as follows:
```
.:53 {
    errors
    nsid CoreDNS
    health
    kubernetes cluster.local in-addr.arpa ip6.arpa {
      pods insecure
      upstream
      fallthrough in-addr.arpa ip6.arpa
    }
    trace zipkin zipkin.kube-system:9411 {
	    every 1000
    }
    prometheus :9153
    forward . /etc/resolv.conf
    cache 30
    reload
    loadbalance
}
```

The `resolv.conf` of the affected CoreDNS pods was:
```
# This file is managed by man:systemd-resolved(8). Do not edit.
#
# This is a dynamic resolv.conf file for connecting local clients directly to
# all known uplink DNS servers. This file lists all configured search domains.
#
# Third party programs must not access this file directly, but only through the
# symlink at /etc/resolv.conf. To manage man:resolv.conf(5) in a different way,
# replace this symlink by a static file or a different symlink.
#
# See man:systemd-resolved.service(8) for details about the supported modes of
# operation for /etc/resolv.conf.

nameserver 10.10.255.252
nameserver 10.10.255.253
search c.my-project.internal google.internal
```

**Comment by fturib** [comment]
Below is a first analysis of the code - I did not run any test to show evidence.
It should be taken as hint and needs to be validated.

Looking into the code, following the stacktrace of panic, I see that:

1- forward/health - NewHealthChecker(protocol) can return nil if the protocol is out of range
2- forward/setup - parseForward(..) the "protocol" map is computed based on the number of parameters provided at the line "forward", but it is then used in loop over range of data that can be bigger than the number of args.
Especially if the param is a file that is substituted with its content - see ParseHostPortOrFile.
=> finally we endup to call NewProxy(..) with a protocol that have the value 0 which is out of range.

In other words, I think the HealthChecker can be nil if the number of lines in /etc/resolv.conf is higher than the number of parameters in the line "forward".

In configuration provided here, that is the case : 1 parameter and 3 lines in resolv.conf.

**Comment by negz** [comment]
@fturib I agree - I noticed the same thing earlier. I've written a test to confirm and am putting together a fix for it now.

**Comment by miekg** [comment]
This is this comment: https://github.com/coredns/coredns/blob/master/plugin/forward/setup.go#L100, didn't anticipate a crash though...

The proper fix is to teach dnsutil.ParseHostOrFile to teach about protocol schemes; `dns://`, `tls://`, etc. 
Some problem probably exist in *proxy*.

/label: bug
/plugin: forward

**Comment by negz** [comment]
@miekg Are you suggesting that the _return signature_ of `dnsutil.ParseHostPortOrFile` be changed to include schemes, as well as updating the parsing logic to respect said schemes?

Given that `plugin/forward` appears to need to know which nameservers are configured to use TLS in order to configure TLS on their respective `*forward.Proxy` instances, I believe such a change would be necessary. Simply changing `dnsutil.ParseHostPortOrFile` to parse schemes correctly and leaving the output unchanged would help choose the correct default port for addresses like `tls://8.8.8.8`, but still require the same kind of hacks as I proposed in #2004.

Assuming you _do_ intend to change the return value of `dnsutil.ParseHostPortOrFile` I agree that is the better fix, but having had a quick scan of the various places it's used I don't have the bandwidth to submit a PR for that right now. I've closed #2004, but am happy to reopen if there's anything useful in there.

**Comment by miekg** [comment]
Yes completely change dnsutil.ParseHostPortOrFile.

**Comment by fturib** [comment]
I think what is not really safe in the code is NewHealthChecker(protocol)  that can return NIL.
I think would be better to return an error if the protocol is invalid.

and force the caller to deal with that error.

**Comment by miekg** [comment]
That too

**Comment by fedya** [comment]
Same here

**Comment by fturib** [comment]
@negz : are you working on this issue ? If yes, are you close to PR or may take longer ?

**Comment by negz** [comment]
@fturib Unfortunately I'm not working on this. I initially said I would, then realised I would not have time to do it properly at the moment. I suspect my comments were read out of order, leading people to believe I was working on this. Sorry!

**Comment by fturib** [comment]
@negz : Thanks for your reply !

**Comment by krzysztof-bronk** [comment]
Is there a workaround for this issue? I've encountered it surprisingly quickly as soon as I upgraded our stack to 1.2.0, and unfortunately had to downgrade for now.

**Comment by miekg** [comment]
This should not be a difficult fix. I'll take a stab tomorrow for a proper
fix.

**Comment by fturib** [comment]
I think the workaround, waiting the fix, avoid using resolv.conf in forward, and rather add the list of nameservers IP directly on the forward line:

"forward . /etc/resolv.conf"  => "forward . 10.10.255.252 10.10.255.253"

**Comment by krzysztof-bronk** [comment]
Ah, I see it now. This could indeed work for us, as our upstreams rarely if ever change.
But I guess I can wait for miek's proper fix now that he got his hands on it ;-)

**Comment by miekg** [comment]
[ Quoting <notifications@github.com> in "Re: [coredns/coredns] plugin/forwar..." ]
>Ah, I see it now. This could indeed work for us, as our upstreams rarely if ever change.
>But I guess I can wait for miek's proper fix now that he got his hands on it ;-)

ack :-) Prolly this week.

**Comment by miekg** [comment]
More work than anticipated: in-flight PR: #2100, not finished yet, because all unit tests fail, but should be easy to fix. I also plan to cleanup *proxy* (maybe in the same PR)

**Comment by miekg** [comment]
Joyful feeling of a net code reduction of 40 lines in PR #2100. Should be good to go, but other than then the unit tests I didn't do any other testing.
